More explicit error messages for new URL(url, base) Constructor

karlcow opened the pull request:
#### e60a65f91bc5b4ae4c6cd886c0a88b510dd8e103
<pre>
More explicit error messages for new URL(url, base) Constructor
<a href="https://bugs.webkit.org/show_bug.cgi?id=256692">https://bugs.webkit.org/show_bug.cgi?id=256692</a>
rdar://109253920

Reviewed by Tim Nguyen.

This gives a more explicit message for `new URL()` instead of just
TypeError. It is also more readable in the console.

* LayoutTests/fast/dom/DOMURL/parsing-expected.txt:
* LayoutTests/fast/dom/DOMURL/url-constructor-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-values/urls/resolve-relative-to-base-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/url/url-constructor.any_exclude=(file_javascript_mailto)-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/url/url-constructor.any.worker_exclude=(file_javascript_mailto)-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/url/url-origin.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/url/url-origin.any.worker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/workers/modules/dedicated-worker-import-referrer-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/workers/modules/shared-worker-import-referrer-expected.txt:
* Source/WebCore/html/DOMURL.cpp:
(WebCore::DOMURL::create):
(WebCore::DOMURL::create):

Canonical link: <a href="https://commits.webkit.org/264129@main">https://commits.webkit.org/264129@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c4e878e811058f1d7c06f782566b999148e55d5c

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/6691 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/6908 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/7090 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/8280 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/6961 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/6689 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/8007 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/6853 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/9857 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/6809 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/7346 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/6124 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/8371 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/4805 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/6054 "Passed tests") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/13882 "1 missing results 120 failures") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/6588 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/6131 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/8856 "run-api-tests-without-change (failure)") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/6622 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/5425 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/6015 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/1608 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/10187 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/6388 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

annevk commented on Source/WebCore/html/DOMURL.cpp:
```diff
@@ -50,15 +50,15 @@ ExceptionOr<Ref<DOMURL>> DOMURL::create(const String& url, const URL& base)
     URL completeURL { base, url };
     if (!completeURL.isValid())
-        return Exception { TypeError };
+        return Exception { TypeError, makeString("URL Constructor: ", url, " is not a valid URL.") };
```
Although our internal API uses `isValid()`, that's not an accurate reflection of what the web standard calls valid. So the web developer facing message should be something like "can't be parsed" here and below.

And perhaps below we should also log the base? As in, "{url} cannot be parsed against {base}."?

karlcow replied:
I'm happy to have more meaningful messages. The more explicit for the Webdev, the better.

karlcow replied:
so there are two different `DOMURL::create`

## `completeURL`

https://github.com/WebKit/WebKit/blob/73de3248842cc21f0547553fda5694324c4ce219/Source/WebCore/html/DOMURL.cpp#L48-L55

This one kicks in for: 

* `new URL("/")`

## `baseURL`

https://github.com/WebKit/WebKit/blob/73de3248842cc21f0547553fda5694324c4ce219/Source/WebCore/html/DOMURL.cpp#L57-L63

this is calling eventually the other `DOMURL::create` if the `baseURL` is not `null` and `valid`.

* `new URL("/", "/")`
* `new URL("/", "")`



Also `canParse` seems like dead code, probably to fix in another PR.

https://github.com/WebKit/WebKit/blob/73de3248842cc21f0547553fda5694324c4ce219/Source/WebCore/html/DOMURL.cpp#L67-L74
https://searchfox.org/wubkat/search?q=symbol:_ZN7WebCore6DOMURL8canParseERKN3WTF6StringES4_&redirect=false

karlcow replied:
canParse was added by @annevk so not dead code.

karlcow replied:
<img width="798" alt="console rendering of error messages" src="https://github.com/WebKit/WebKit/assets/505230/fd4365bf-a830-4a2e-bc96-7ce48306f892">

This is the result of my modifications, I will adjust the tests.

annevk reviewed (commented):
The wording seems fine, but I'm a little unsure about repeating the constructor in the message. Is that normal in exceptions? I would have expected something more abstract, e.g.,

> "blah" cannot be parsed as URL.

annevk commented on LayoutTests/imported/w3c/web-platform-tests/workers/modules/dedicated-worker-import-referrer-expected.txt:
```diff
@@ -11,13 +11,13 @@ PASS Same-origin top-level module script loading with "same-origin" referrer pol
-FAIL Cross-origin dynamic import with "no-referrer" referrer policy. promise_test: Unhandled rejection with value: object "TypeError: Type error"
-FAIL Cross-origin dynamic import with "origin" referrer policy. promise_test: Unhandled rejection with value: object "TypeError: Type error"
-FAIL Cross-origin dynamic import with "same-origin" referrer policy. promise_test: Unhandled rejection with value: object "TypeError: Type error"
+FAIL Cross-origin dynamic import with "no-referrer" referrer policy. promise_test: Unhandled rejection with value: object "TypeError: "Import failed: TypeError: Cross-origin script load denied by Cross-Origin Resource Sharing policy." cannot be parsed as a URL."
```
This looks weird?

karlcow replied:
Yes it looks funny. 
I don't know why the value of URL is this in that test. 
@nt1m do you know why this is happening.

karlcow replied:
This should probably be investigated, Maybe something about the test framework.

annevk commented on LayoutTests/fast/dom/DOMURL/parsing-expected.txt:
```diff
@@ -13,7 +13,7 @@ PASS breakDownURL('file:///Users/darin') is 'protocol=file:, pathname=/Users/dar
 PASS breakDownURL('http://a:b@c:1/e/f?g%h') is 'protocol=http:, username=a, password=b, hostname=c, host=c:1, port=1, pathname=/e/f, search=?g%h, origin=http://c:1'
 PASS breakDownURL('http://ex%61mple.com/') is 'protocol=http:, host=example.com, pathname=/, origin=http://example.com, toString=http://example.com/'
-PASS breakDownURL('http://ex%2fmple.com/') threw exception TypeError: Type error.
+PASS breakDownURL('http://ex%2fmple.com/') threw exception TypeError: "http://ex%2fmple.com/" cannot be parsed as a URL..
```
The double dot looks a bit funny here, but I guess that's normal?

karlcow replied:
the first dot comes from the error message, and probably the second dot comes from the test. Let see what is the test. 

The second dot is coming from the test framework

https://github.com/WebKit/WebKit/blob/22a32809f2d43aa8c6984784fe2a53078e43a47f/LayoutTests/resources/js-test.js#L709

karlcow replied:
Does that mean, error messages should not have dots.

karlcow replied:
I will merge as-is.

annevk reviewed (commented):
This looks good to me, though the cross-origin dynamic import test results seem suspect. Some kind of nested exception? Not a reviewer so no r+.

annevk reviewed (approved):
Approving to remove my "request changes". Again, not a WebKit reviewer. 😊

karlcow committed:
More explicit error messages for new URL(url, base) Constructor
https://bugs.webkit.org/show_bug.cgi?id=256692
rdar://109253920

Reviewed by Tim Nguyen.

This gives a more explicit message for `new URL()` instead of just
TypeError. It is also more readable in the console.

* LayoutTests/fast/dom/DOMURL/parsing-expected.txt:
* LayoutTests/fast/dom/DOMURL/url-constructor-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-values/urls/resolve-relative-to-base-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/url/url-constructor.any_exclude=(file_javascript_mailto)-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/url/url-constructor.any.worker_exclude=(file_javascript_mailto)-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/url/url-origin.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/url/url-origin.any.worker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/workers/modules/dedicated-worker-import-referrer-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/workers/modules/shared-worker-import-referrer-expected.txt:
* Source/WebCore/html/DOMURL.cpp:
(WebCore::DOMURL::create):
(WebCore::DOMURL::create):

Canonical link: https://commits.webkit.org/264129@main
